{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fasta import FASTQ\n",
    "from Bio import AlignIO\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters for input and output\n",
    "min_seq_len_to_report = 3\n",
    "skip_rows = 1\n",
    "set_destinated_protein_id_for_chimerax = \"8VYN_1|Chains\"\n",
    "pdb_offset = 1\n",
    "# List for plotting in ChimeraX for conservative regions\n",
    "color_list = [\"pink\", \"orange\", \"yellow\", \"white\", \"purple\", \"brown\", \"cyan\",\\\n",
    "              \"lime\", \"gray\", \"dark slate blue\", \"turquoise\", \"gold\", \"tan\", \"chocolate\", \"tomato\", \"violet\",\n",
    "              \"indigo\", \"navy\", \"deep pink\", \"chartreuse\", \"spring green\", \"goldenrod\", \"black\", \"sky blue\"]\n",
    "\n",
    "color_list = [\"yellow\"] # Conserved\n",
    "chain_color = \"dark grey\"  # Non-Conserved\n",
    "\n",
    "# Dimer, Trimer? or even more?\n",
    "chain = \"A, B, C\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do part 1, run only at first time to do alignment.\n",
    "try: \n",
    "    alignment = AlignIO.read(\"2_aligned.aln-clustal_num\", \"clustal\")\n",
    "    gen_align = False\n",
    "except:\n",
    "    gen_align = True\n",
    "\n",
    "if gen_align == True:\n",
    "    !clustalo -i 1_sequences.fasta --outfmt=clu --wrap=2048 --outfile=2_aligned.aln-clustal_num --force --threads=16;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First Protein is: 8VYN_1|Chains\n"
     ]
    }
   ],
   "source": [
    "# Read aligned data, missing part show with \"-\"\n",
    "pdb_offset = pdb_offset - 1\n",
    "alignment = AlignIO.read(\"2_aligned.aln-clustal_num\", \"clustal\")\n",
    "pre_seq_offset_in_clustal_file = 16 # Could be 16, 17, 20, 24, you are right when Length of sequence = After cut\n",
    "\n",
    "try:\n",
    "    shutil.copy(\"2_aligned.aln-clustal_num\", \"3_aligned.aln-clustal_num.txt\")\n",
    "    f = open(\"3_aligned.aln-clustal_num.txt\", \"r\")\n",
    "except:\n",
    "    print(\"Failed to open file.\")\n",
    "\n",
    "found_offset = False\n",
    "loop_count = 0\n",
    "i = 0\n",
    "\n",
    "for x in f:\n",
    "    i = i + 1\n",
    "\n",
    "    if len(set_destinated_protein_id_for_chimerax) > 1:\n",
    "        destinated_protein_id_for_chimerax = set_destinated_protein_id_for_chimerax\n",
    "        break\n",
    "\n",
    "    if i > skip_rows and np.size([*x]) > 1:\n",
    "        destinated_protein_id_for_chimerax = \"\".join([*x][0:x.find(\".\") + 2])\n",
    "        break\n",
    "\n",
    "print(\"First Protein is:\", destinated_protein_id_for_chimerax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Offset: 19\n",
      "Ref_seq_Id 8VYN_1|Chains\n",
      "Sequence ['M', 'E', 'S', 'R', 'I', 'W', 'C', 'L', 'V', 'V', 'C', 'V', 'N', 'L', 'C', 'I', 'V', 'C', 'L', 'G', 'A', 'A', 'V', 'S', 'S', 'S', 'S', 'T', 'R', 'G', 'T', 'S', 'A', 'T', 'H', 'S', 'H', 'H', 'S', 'S', 'H', 'T', 'T', 'S', 'A', 'A', 'H', 'S', 'R', 'S', 'G', 'S', 'V', 'S', 'Q', 'R', 'V', 'T', 'S', 'S', 'Q', 'T', 'V', 'S', 'H', 'G', 'V', 'N', 'E', 'T', 'I', 'Y', 'N', 'T', 'T', 'L', 'K', 'Y', 'G', 'D', 'V', 'V', 'G', 'V', 'N', 'T', 'T', 'K', 'Y', 'P', 'Y', 'R', 'V', 'C', 'S', 'M', 'A', 'Q', 'G', 'L', 'D', 'L', 'I', 'R', 'F', 'E', 'R', 'N', 'I', 'V', 'C', 'T', 'S', 'M', 'K', 'P', 'I', 'N', 'E', 'D', 'L', 'D', 'E', 'G', 'I', 'M', 'V', 'V', 'Y', 'K', 'R', 'N', 'I', 'C', 'A', 'H', 'T', 'F', 'K', 'V', 'R', 'V', 'Y', 'Q', 'K', 'V', 'L', 'T', 'F', 'R', 'R', 'S', 'Y', 'A', 'Y', 'I', 'H', 'T', 'T', 'Y', 'L', 'L', 'G', 'S', 'N', 'T', 'E', 'Y', 'V', 'A', 'P', 'P', 'M', 'W', 'E', 'I', 'H', 'H', 'I', 'N', 'S', 'H', 'S', 'Q', 'C', 'Y', 'S', 'S', 'Y', 'S', 'R', 'V', 'I', 'A', 'G', 'T', 'V', 'F', 'V', 'A', 'Y', 'H', 'R', 'D', 'S', 'Y', 'E', 'N', 'K', 'T', 'M', 'Q', 'L', 'M', 'P', 'D', 'D', 'Y', 'S', 'N', 'T', 'C', 'S', 'T', 'R', 'Y', 'V', 'T', 'V', 'K', 'D', 'Q', 'W', 'H', 'S', 'R', 'G', 'S', 'T', 'W', 'L', 'Y', 'R', 'E', 'T', 'S', 'N', 'L', 'N', 'C', 'M', 'V', 'T', 'I', 'T', 'T', 'A', 'R', 'S', 'K', 'Y', 'P', 'Y', 'H', 'F', 'F', 'I', 'T', 'S', 'T', 'G', 'D', 'V', 'V', 'D', 'I', 'S', 'P', 'F', 'Y', 'N', 'G', 'T', 'N', 'R', 'N', 'A', 'S', 'Y', 'F', 'G', 'E', 'N', 'A', 'D', 'K', 'F', 'F', 'I', 'F', 'P', 'N', 'Y', 'T', 'I', 'V', 'S', 'D', 'F', 'G', 'R', 'P', 'N', 'S', 'A', 'L', 'E', 'T', 'H', 'R', 'L', 'V', 'A', 'F', 'L', 'E', 'R', 'A', 'D', 'S', 'V', 'I', 'S', 'W', 'D', 'I', 'Q', 'D', 'E', 'K', 'N', 'V', 'T', 'C', 'Q', 'L', 'T', 'F', 'W', 'E', 'A', 'S', 'E', 'R', 'T', 'I', 'R', 'S', 'E', 'A', 'E', 'D', 'S', 'Y', 'H', 'F', 'S', 'S', 'A', 'K', 'M', 'T', 'A', 'T', 'F', 'L', 'S', 'K', 'K', 'Q', 'E', 'V', 'N', 'M', 'S', 'D', 'S', 'A', 'L', 'D', 'C', 'V', 'R', 'D', 'E', 'A', 'I', 'N', 'K', 'L', 'Q', 'Q', 'I', 'F', 'N', 'T', 'S', 'Y', 'N', 'Q', 'T', 'Y', 'E', 'K', 'Y', 'G', 'N', 'V', 'S', 'V', 'F', 'E', 'T', 'T', 'G', 'G', 'L', 'V', 'V', 'F', 'W', 'Q', 'G', 'I', 'K', 'Q', 'K', 'S', 'L', 'V', 'E', 'L', 'E', 'R', 'L', 'A', 'N', 'R', 'S', 'S', 'L', 'N', 'L', 'T', 'H', 'N', 'S', 'T', 'K', 'S', 'S', 'T', 'D', 'G', 'N', 'N', 'A', 'T', 'H', 'L', 'S', 'N', 'M', 'E', 'S', 'V', 'H', 'N', 'L', 'V', 'Y', 'A', 'Q', 'L', 'Q', 'F', 'T', 'Y', 'D', 'T', 'L', 'R', 'G', 'Y', 'I', 'N', 'R', 'A', 'L', 'A', 'Q', 'I', 'A', 'E', 'A', 'W', 'C', 'V', 'D', 'Q', 'R', 'R', 'T', 'L', 'E', 'V', 'F', 'K', 'E', 'L', 'S', 'K', 'I', 'N', 'P', 'S', 'A', 'I', 'L', 'S', 'A', 'I', 'Y', 'N', 'K', 'P', 'I', 'A', 'A', 'R', 'F', 'M', 'G', 'D', 'V', 'L', 'G', 'L', 'A', 'S', 'C', 'V', 'T', 'I', 'N', 'Q', 'T', 'S', 'V', 'K', 'V', 'L', 'R', 'D', 'M', 'N', 'V', 'K', 'E', 'S', 'P', 'G', 'R', 'C', 'Y', 'S', 'R', 'P', 'V', 'V', 'I', 'F', 'N', 'F', 'A', 'N', 'S', 'S', 'Y', 'V', 'Q', 'Y', 'G', 'Q', 'L', 'G', 'E', 'D', 'N', 'E', 'I', 'L', 'L', 'G', 'N', 'H', 'R', 'T', 'E', 'E', 'C', 'Q', 'L', 'P', 'S', 'L', 'K', 'I', 'F', 'I', 'A', 'G', 'N', 'S', 'A', 'Y', 'E', 'Y', 'V', 'D', 'Y', 'L', 'F', 'K', 'R', 'M', 'I', 'D', 'L', 'S', 'S', 'I', 'S', 'T', 'V', 'D', 'S', 'M', 'I', 'A', 'L', 'D', 'C', 'D', 'P', 'L', 'C', 'N', 'T', 'D', 'F', 'R', 'V', 'L', 'E', 'L', 'Y', 'S', 'Q', 'K', 'E', 'L', 'R', 'S', 'S', 'N', 'V', 'F', 'D', 'L', 'E', 'E', 'I', 'M', 'R', 'E', 'F', 'N', 'S', 'Y', 'K', 'Q', 'R', 'V', 'K', 'Y', 'V', 'E', 'D', 'K', 'V', 'V', 'D', 'P', 'G', 'S', 'G', 'Y', 'I', 'P', 'E', 'A', 'P', 'R', 'D', 'G', 'Q', 'A', 'Y', 'V', 'R', 'K', 'D', 'G', 'E', 'W', 'V', 'L', 'L', 'S', 'T', 'F', 'L', 'G', 'A', 'A', 'A', 'S', 'L', 'E', 'V', 'L', 'F', 'Q', 'G', 'P', 'G', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'S', 'A', 'W', 'S', 'H', 'P', 'Q', 'F', 'E', 'K', 'G', 'G', 'A', 'S', 'G', 'G', 'G', 'G', 'S', 'G', 'G', 'S', 'A', 'W', 'S', 'H', 'P', 'Q', 'F', 'E', 'K']\n",
      "Length of sequence: 786\n",
      "Before cut 786\n",
      "After cut 786\n"
     ]
    }
   ],
   "source": [
    "# Read aligned data, missing part show with \"-\"\n",
    "alignment = AlignIO.read(\"2_aligned.aln-clustal_num\", \"clustal\")\n",
    "pre_seq_offset_in_clustal_file = 16 # Could be 16, 17, 20, 24, you are right when Length of sequence = After cut\n",
    "f = open(\"3_aligned.aln-clustal_num.txt\", \"r\")\n",
    "\n",
    "found_offset = False\n",
    "loop_count = 0\n",
    "\n",
    "while found_offset == False:\n",
    "    conservative_seq = \"\"\n",
    "    i = 0\n",
    "    loop_count += 1\n",
    "    alignment = AlignIO.read(\"2_aligned.aln-clustal_num\", \"clustal\")\n",
    "    f = open(\"3_aligned.aln-clustal_num.txt\", \"r\")\n",
    "\n",
    "    for x in f:\n",
    "        i = i + 1\n",
    "        if i >= skip_rows:\n",
    "            if x[0:pre_seq_offset_in_clustal_file] == \" \" * pre_seq_offset_in_clustal_file:\n",
    "                conservative_seq += x[pre_seq_offset_in_clustal_file:]\n",
    "                #print(conservative_seq)\n",
    "        \n",
    "    conservative_seq = conservative_seq.replace('\\n',\"\")\n",
    "    conservative_seq = [*conservative_seq]\n",
    "\n",
    "    for record in alignment:\n",
    "        if record.id == destinated_protein_id_for_chimerax:\n",
    "            #print(record.id)\n",
    "            sequence = str(record.seq)\n",
    "            seq_replaced = [*sequence.replace(\"-\", \"\")]\n",
    "            #print(sequence)\n",
    "            sequence = np.array([*sequence])\n",
    "            rm_index = np.where(sequence == \"-\")[0]\n",
    "            #print(\"Length of sequence:\", len(seq_replaced))\n",
    "\n",
    "    rm_index = rm_index[::-1]\n",
    "    #print(\"Before cut\", len(conservative_seq))\n",
    "\n",
    "    for j in rm_index:\n",
    "        del conservative_seq[int(j)]\n",
    "\n",
    "    #print(\"After cut\", len(conservative_seq))\n",
    "\n",
    "    seq_len = len(seq_replaced)\n",
    "    cut_end = len(conservative_seq)\n",
    "\n",
    "    if cut_end == 0 or seq_len == 0:\n",
    "        pre_seq_offset_in_clustal_file = pre_seq_offset_in_clustal_file - 1\n",
    "\n",
    "    \n",
    "    elif seq_len == cut_end:\n",
    "        found_offset = True\n",
    "        print(\"Found Offset:\", pre_seq_offset_in_clustal_file)\n",
    "        print(\"Ref_seq_Id\", destinated_protein_id_for_chimerax)\n",
    "        print(\"Sequence\", seq_replaced)\n",
    "        print(\"Length of sequence:\", len(seq_replaced))\n",
    "        print(\"Before cut\", len(conservative_seq))\n",
    "        print(\"After cut\", len(conservative_seq))\n",
    "\n",
    "    elif seq_len < cut_end:\n",
    "        pre_seq_offset_in_clustal_file = pre_seq_offset_in_clustal_file + 1\n",
    "\n",
    "    if loop_count > 32:\n",
    "        raise(\"FileExistsError\", \"Alignment file is wrong or corrpted\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index 1 M  \n",
      "Index 2 E  \n",
      "Index 3 S  \n",
      "Index 4 R  \n",
      "Index 5 I  \n",
      "Index 6 W  \n",
      "Index 7 C  \n",
      "Index 8 L  \n",
      "Index 9 V  \n",
      "Index 10 V  \n",
      "Index 11 C  \n",
      "Index 12 V  \n",
      "Index 13 N  \n",
      "Index 14 L  \n",
      "Index 15 C  \n",
      "Index 16 I  \n",
      "Index 17 V  \n",
      "Index 18 C  \n",
      "Index 19 L  \n",
      "Index 20 G  \n",
      "Index 21 A  \n",
      "Index 22 A  \n",
      "Index 23 V  \n",
      "Index 24 S  \n",
      "Index 25 S  \n",
      "Index 26 S  \n",
      "Index 27 S  \n",
      "Index 28 T  \n",
      "Index 29 R  \n",
      "Index 30 G  \n",
      "Index 31 T  \n",
      "Index 32 S  \n",
      "Index 33 A  \n",
      "Index 34 T  \n",
      "Index 35 H  \n",
      "Index 36 S  \n",
      "Index 37 H  \n",
      "Index 38 H  \n",
      "Index 39 S  \n",
      "Index 40 S  \n",
      "Index 41 H  \n",
      "Index 42 T  \n",
      "Index 43 T  \n",
      "Index 44 S  \n",
      "Index 45 A  \n",
      "Index 46 A  \n",
      "Index 47 H  \n",
      "Index 48 S  \n",
      "Index 49 R  \n",
      "Index 50 S  \n",
      "Index 51 G  \n",
      "Index 52 S  \n",
      "Index 53 V  \n",
      "Index 54 S  \n",
      "Index 55 Q  \n",
      "Index 56 R  \n",
      "Index 57 V  \n",
      "Index 58 T  \n",
      "Index 59 S  \n",
      "Index 60 S  \n",
      "Index 61 Q  \n",
      "Index 62 T  \n",
      "Index 63 V  \n",
      "Index 64 S  \n",
      "Index 65 H  \n",
      "Index 66 G  \n",
      "Index 67 V  \n",
      "Index 68 N  \n",
      "Index 69 E  \n",
      "Index 70 T  \n",
      "Index 71 I  \n",
      "Index 72 Y  \n",
      "Index 73 N  \n",
      "Index 74 T  \n",
      "Index 75 T  \n",
      "Index 76 L  \n",
      "Index 77 K  \n",
      "Index 78 Y  \n",
      "Index 79 G  \n",
      "Index 80 D  \n",
      "Index 81 V  \n",
      "Index 82 V  \n",
      "Index 83 G  \n",
      "Index 84 V  \n",
      "Index 85 N  \n",
      "Index 86 T  \n",
      "Index 87 T  \n",
      "Index 88 K  \n",
      "Index 89 Y  \n",
      "Index 90 P  \n",
      "Index 91 Y  \n",
      "Index 92 R  \n",
      "Index 93 V  \n",
      "Index 94 C  \n",
      "Index 95 S  \n",
      "Index 96 M  \n",
      "Index 97 A  \n",
      "Index 98 Q  \n",
      "Index 99 G  \n",
      "Index 100 L  \n",
      "Index 101 D  \n",
      "Index 102 L  \n",
      "Index 103 I  \n",
      "Index 104 R  \n",
      "Index 105 F  \n",
      "Index 106 E  \n",
      "Index 107 R  \n",
      "Index 108 N  \n",
      "Index 109 I  \n",
      "Index 110 V  \n",
      "Index 111 C  \n",
      "Index 112 T  \n",
      "Index 113 S  \n",
      "Index 114 M  \n",
      "Index 115 K  \n",
      "Index 116 P  \n",
      "Index 117 I  \n",
      "Index 118 N  \n",
      "Index 119 E  \n",
      "Index 120 D  \n",
      "Index 121 L  \n",
      "Index 122 D  \n",
      "Index 123 E  \n",
      "Index 124 G  \n",
      "Index 125 I  \n",
      "Index 126 M  \n",
      "Index 127 V  \n",
      "Index 128 V  \n",
      "Index 129 Y  \n",
      "Index 130 K  \n",
      "Index 131 R  \n",
      "Index 132 N  \n",
      "Index 133 I  \n",
      "Index 134 C  \n",
      "Index 135 A  \n",
      "Index 136 H  \n",
      "Index 137 T  \n",
      "Index 138 F  \n",
      "Index 139 K  \n",
      "Index 140 V  \n",
      "Index 141 R  \n",
      "Index 142 V  \n",
      "Index 143 Y  \n",
      "Index 144 Q  \n",
      "Index 145 K  \n",
      "Index 146 V  \n",
      "Index 147 L  \n",
      "Index 148 T  \n",
      "Index 149 F  \n",
      "Index 150 R  \n",
      "Index 151 R  \n",
      "Index 152 S  \n",
      "Index 153 Y  \n",
      "Index 154 A  \n",
      "Index 155 Y  \n",
      "Index 156 I  \n",
      "Index 157 H  \n",
      "Index 158 T  \n",
      "Index 159 T  \n",
      "Index 160 Y  \n",
      "Index 161 L  \n",
      "Index 162 L  \n",
      "Index 163 G  \n",
      "Index 164 S  \n",
      "Index 165 N  \n",
      "Index 166 T  \n",
      "Index 167 E  \n",
      "Index 168 Y  \n",
      "Index 169 V  \n",
      "Index 170 A  \n",
      "Index 171 P  \n",
      "Index 172 P  \n",
      "Index 173 M  \n",
      "Index 174 W  \n",
      "Index 175 E  \n",
      "Index 176 I  \n",
      "Index 177 H  \n",
      "Index 178 H  \n",
      "Index 179 I  \n",
      "Index 180 N  \n",
      "Index 181 S  \n",
      "Index 182 H  \n",
      "Index 183 S  \n",
      "Index 184 Q  \n",
      "Index 185 C  \n",
      "Index 186 Y  \n",
      "Index 187 S  \n",
      "Index 188 S  \n",
      "Index 189 Y  \n",
      "Index 190 S  \n",
      "Index 191 R  \n",
      "Index 192 V  \n",
      "Index 193 I  \n",
      "Index 194 A  \n",
      "Index 195 G  \n",
      "Index 196 T  \n",
      "Index 197 V  \n",
      "Index 198 F  \n",
      "Index 199 V  \n",
      "Index 200 A  \n",
      "Index 201 Y  \n",
      "Index 202 H  \n",
      "Index 203 R  \n",
      "Index 204 D  \n",
      "Index 205 S  \n",
      "Index 206 Y  \n",
      "Index 207 E  \n",
      "Index 208 N  \n",
      "Index 209 K  \n",
      "Index 210 T  \n",
      "Index 211 M  \n",
      "Index 212 Q  \n",
      "Index 213 L  \n",
      "Index 214 M  \n",
      "Index 215 P  \n",
      "Index 216 D  \n",
      "Index 217 D  \n",
      "Index 218 Y  \n",
      "Index 219 S  \n",
      "Index 220 N  \n",
      "Index 221 T  \n",
      "Index 222 C  \n",
      "Index 223 S  \n",
      "Index 224 T  \n",
      "Index 225 R  \n",
      "Index 226 Y  \n",
      "Index 227 V  \n",
      "Index 228 T  \n",
      "Index 229 V  \n",
      "Index 230 K  \n",
      "Index 231 D  \n",
      "Index 232 Q  \n",
      "Index 233 W  \n",
      "Index 234 H  \n",
      "Index 235 S  \n",
      "Index 236 R  \n",
      "Index 237 G  \n",
      "Index 238 S  \n",
      "Index 239 T  \n",
      "Index 240 W  \n",
      "Index 241 L  \n",
      "Index 242 Y  \n",
      "Index 243 R  \n",
      "Index 244 E  \n",
      "Index 245 T  \n",
      "Index 246 S  \n",
      "Index 247 N  \n",
      "Index 248 L  \n",
      "Index 249 N  \n",
      "Index 250 C  \n",
      "Index 251 M  \n",
      "Index 252 V  \n",
      "Index 253 T  \n",
      "Index 254 I  \n",
      "Index 255 T  \n",
      "Index 256 T  \n",
      "Index 257 A  \n",
      "Index 258 R  \n",
      "Index 259 S  \n",
      "Index 260 K  \n",
      "Index 261 Y  \n",
      "Index 262 P  \n",
      "Index 263 Y  \n",
      "Index 264 H  \n",
      "Index 265 F  \n",
      "Index 266 F  \n",
      "Index 267 I  \n",
      "Index 268 T  \n",
      "Index 269 S  \n",
      "Index 270 T  \n",
      "Index 271 G  \n",
      "Index 272 D  \n",
      "Index 273 V  \n",
      "Index 274 V  \n",
      "Index 275 D  \n",
      "Index 276 I  \n",
      "Index 277 S  \n",
      "Index 278 P  \n",
      "Index 279 F  \n",
      "Index 280 Y  \n",
      "Index 281 N  \n",
      "Index 282 G  \n",
      "Index 283 T  \n",
      "Index 284 N  \n",
      "Index 285 R  \n",
      "Index 286 N  \n",
      "Index 287 A  \n",
      "Index 288 S  \n",
      "Index 289 Y  \n",
      "Index 290 F  \n",
      "Index 291 G  \n",
      "Index 292 E  \n",
      "Index 293 N  \n",
      "Index 294 A  \n",
      "Index 295 D  \n",
      "Index 296 K  \n",
      "Index 297 F  \n",
      "Index 298 F  \n",
      "Index 299 I  \n",
      "Index 300 F  \n",
      "Index 301 P  \n",
      "Index 302 N  \n",
      "Index 303 Y  \n",
      "Index 304 T  \n",
      "Index 305 I  \n",
      "Index 306 V  \n",
      "Index 307 S  \n",
      "Index 308 D  \n",
      "Index 309 F  \n",
      "Index 310 G  \n",
      "Index 311 R  \n",
      "Index 312 P  \n",
      "Index 313 N  \n",
      "Index 314 S  \n",
      "Index 315 A  \n",
      "Index 316 L  \n",
      "Index 317 E  \n",
      "Index 318 T  \n",
      "Index 319 H  \n",
      "Index 320 R  \n",
      "Index 321 L  \n",
      "Index 322 V  \n",
      "Index 323 A  \n",
      "Index 324 F  \n",
      "Index 325 L  \n",
      "Index 326 E  \n",
      "Index 327 R  \n",
      "Index 328 A  \n",
      "Index 329 D  \n",
      "Index 330 S  \n",
      "Index 331 V  \n",
      "Index 332 I  \n",
      "Index 333 S  \n",
      "Index 334 W  \n",
      "Index 335 D  \n",
      "Index 336 I  \n",
      "Index 337 Q  \n",
      "Index 338 D  \n",
      "Index 339 E  \n",
      "Index 340 K  \n",
      "Index 341 N  \n",
      "Index 342 V  \n",
      "Index 343 T  \n",
      "Index 344 C  \n",
      "Index 345 Q  \n",
      "Index 346 L  \n",
      "Index 347 T  \n",
      "Index 348 F  \n",
      "Index 349 W  \n",
      "Index 350 E  \n",
      "Index 351 A  \n",
      "Index 352 S  \n",
      "Index 353 E  \n",
      "Index 354 R  \n",
      "Index 355 T  \n",
      "Index 356 I  \n",
      "Index 357 R  \n",
      "Index 358 S  \n",
      "Index 359 E  \n",
      "Index 360 A  \n",
      "Index 361 E  \n",
      "Index 362 D  \n",
      "Index 363 S  \n",
      "Index 364 Y  \n",
      "Index 365 H  \n",
      "Index 366 F  \n",
      "Index 367 S  \n",
      "Index 368 S  \n",
      "Index 369 A  \n",
      "Index 370 K  \n",
      "Index 371 M  \n",
      "Index 372 T  \n",
      "Index 373 A  \n",
      "Index 374 T  \n",
      "Index 375 F  \n",
      "Index 376 L  \n",
      "Index 377 S  \n",
      "Index 378 K  \n",
      "Index 379 K  \n",
      "Index 380 Q  \n",
      "Index 381 E  \n",
      "Index 382 V  \n",
      "Index 383 N  \n",
      "Index 384 M  \n",
      "Index 385 S  \n",
      "Index 386 D  \n",
      "Index 387 S  \n",
      "Index 388 A  \n",
      "Index 389 L  \n",
      "Index 390 D  \n",
      "Index 391 C  \n",
      "Index 392 V  \n",
      "Index 393 R  \n",
      "Index 394 D  \n",
      "Index 395 E  \n",
      "Index 396 A  \n",
      "Index 397 I  \n",
      "Index 398 N  \n",
      "Index 399 K  \n",
      "Index 400 L  \n",
      "Index 401 Q  \n",
      "Index 402 Q  \n",
      "Index 403 I  \n",
      "Index 404 F  \n",
      "Index 405 N  \n",
      "Index 406 T  \n",
      "Index 407 S  \n",
      "Index 408 Y  \n",
      "Index 409 N  \n",
      "Index 410 Q  \n",
      "Index 411 T  \n",
      "Index 412 Y  \n",
      "Index 413 E  \n",
      "Index 414 K  \n",
      "Index 415 Y  \n",
      "Index 416 G  \n",
      "Index 417 N  \n",
      "Index 418 V  \n",
      "Index 419 S  \n",
      "Index 420 V  \n",
      "Index 421 F  \n",
      "Index 422 E  \n",
      "Index 423 T  \n",
      "Index 424 T  \n",
      "Index 425 G  \n",
      "Index 426 G  \n",
      "Index 427 L  \n",
      "Index 428 V  \n",
      "Index 429 V  \n",
      "Index 430 F  \n",
      "Index 431 W  \n",
      "Index 432 Q  \n",
      "Index 433 G  \n",
      "Index 434 I  \n",
      "Index 435 K  \n",
      "Index 436 Q  \n",
      "Index 437 K  \n",
      "Index 438 S  \n",
      "Index 439 L  \n",
      "Index 440 V  \n",
      "Index 441 E  \n",
      "Index 442 L  \n",
      "Index 443 E  \n",
      "Index 444 R  \n",
      "Index 445 L  \n",
      "Index 446 A  \n",
      "Index 447 N  \n",
      "Index 448 R  \n",
      "Index 449 S  \n",
      "Index 450 S  \n",
      "Index 451 L  \n",
      "Index 452 N  \n",
      "Index 453 L  \n",
      "Index 454 T  \n",
      "Index 455 H  \n",
      "Index 456 N  \n",
      "Index 457 S  \n",
      "Index 458 T  \n",
      "Index 459 K  \n",
      "Index 460 S  \n",
      "Index 461 S  \n",
      "Index 462 T  \n",
      "Index 463 D  \n",
      "Index 464 G  \n",
      "Index 465 N  \n",
      "Index 466 N  \n",
      "Index 467 A  \n",
      "Index 468 T  \n",
      "Index 469 H  \n",
      "Index 470 L  \n",
      "Index 471 S  \n",
      "Index 472 N  \n",
      "Index 473 M  \n",
      "Index 474 E  \n",
      "Index 475 S  \n",
      "Index 476 V  \n",
      "Index 477 H  \n",
      "Index 478 N  \n",
      "Index 479 L  \n",
      "Index 480 V  \n",
      "Index 481 Y  \n",
      "Index 482 A  \n",
      "Index 483 Q  \n",
      "Index 484 L  \n",
      "Index 485 Q  \n",
      "Index 486 F  \n",
      "Index 487 T  \n",
      "Index 488 Y  \n",
      "Index 489 D  \n",
      "Index 490 T  \n",
      "Index 491 L  \n",
      "Index 492 R  \n",
      "Index 493 G  \n",
      "Index 494 Y  \n",
      "Index 495 I  \n",
      "Index 496 N  \n",
      "Index 497 R  \n",
      "Index 498 A  \n",
      "Index 499 L  \n",
      "Index 500 A  \n",
      "Index 501 Q  \n",
      "Index 502 I  \n",
      "Index 503 A  \n",
      "Index 504 E  \n",
      "Index 505 A  \n",
      "Index 506 W  \n",
      "Index 507 C  \n",
      "Index 508 V  \n",
      "Index 509 D  \n",
      "Index 510 Q  \n",
      "Index 511 R  \n",
      "Index 512 R  \n",
      "Index 513 T  \n",
      "Index 514 L  \n",
      "Index 515 E  \n",
      "Index 516 V  \n",
      "Index 517 F  \n",
      "Index 518 K  \n",
      "Index 519 E  \n",
      "Index 520 L  \n",
      "Index 521 S  \n",
      "Index 522 K  \n",
      "Index 523 I  \n",
      "Index 524 N  \n",
      "Index 525 P  \n",
      "Index 526 S  \n",
      "Index 527 A  \n",
      "Index 528 I  \n",
      "Index 529 L  \n",
      "Index 530 S  \n",
      "Index 531 A  \n",
      "Index 532 I  \n",
      "Index 533 Y  \n",
      "Index 534 N  \n",
      "Index 535 K  \n",
      "Index 536 P  \n",
      "Index 537 I  \n",
      "Index 538 A  \n",
      "Index 539 A  \n",
      "Index 540 R  \n",
      "Index 541 F  \n",
      "Index 542 M  \n",
      "Index 543 G  \n",
      "Index 544 D  \n",
      "Index 545 V  \n",
      "Index 546 L  \n",
      "Index 547 G  \n",
      "Index 548 L  \n",
      "Index 549 A  \n",
      "Index 550 S  \n",
      "Index 551 C  \n",
      "Index 552 V  \n",
      "Index 553 T  \n",
      "Index 554 I  \n",
      "Index 555 N  \n",
      "Index 556 Q  \n",
      "Index 557 T  \n",
      "Index 558 S  \n",
      "Index 559 V  \n",
      "Index 560 K  \n",
      "Index 561 V  \n",
      "Index 562 L  \n",
      "Index 563 R  \n",
      "Index 564 D  \n",
      "Index 565 M  \n",
      "Index 566 N  \n",
      "Index 567 V  \n",
      "Index 568 K  \n",
      "Index 569 E  \n",
      "Index 570 S  \n",
      "Index 571 P  \n",
      "Index 572 G  \n",
      "Index 573 R  \n",
      "Index 574 C  \n",
      "Index 575 Y  \n",
      "Index 576 S  \n",
      "Index 577 R  \n",
      "Index 578 P  \n",
      "Index 579 V  \n",
      "Index 580 V  \n",
      "Index 581 I  \n",
      "Index 582 F  \n",
      "Index 583 N  \n",
      "Index 584 F  \n",
      "Index 585 A  \n",
      "Index 586 N  \n",
      "Index 587 S  \n",
      "Index 588 S  \n",
      "Index 589 Y  \n",
      "Index 590 V  \n",
      "Index 591 Q  \n",
      "Index 592 Y  \n",
      "Index 593 G  \n",
      "Index 594 Q  \n",
      "Index 595 L  \n",
      "Index 596 G  \n",
      "Index 597 E  \n",
      "Index 598 D  \n",
      "Index 599 N  \n",
      "Index 600 E  \n",
      "Index 601 I  \n",
      "Index 602 L  \n",
      "Index 603 L  \n",
      "Index 604 G  \n",
      "Index 605 N  \n",
      "Index 606 H  \n",
      "Index 607 R  \n",
      "Index 608 T  \n",
      "Index 609 E  \n",
      "Index 610 E  \n",
      "Index 611 C  \n",
      "Index 612 Q  \n",
      "Index 613 L  \n",
      "Index 614 P  \n",
      "Index 615 S  \n",
      "Index 616 L  \n",
      "Index 617 K  \n",
      "Index 618 I  \n",
      "Index 619 F  \n",
      "Index 620 I  \n",
      "Index 621 A  \n",
      "Index 622 G  \n",
      "Index 623 N  \n",
      "Index 624 S  \n",
      "Index 625 A  \n",
      "Index 626 Y  \n",
      "Index 627 E  \n",
      "Index 628 Y  \n",
      "Index 629 V  \n",
      "Index 630 D  \n",
      "Index 631 Y  \n",
      "Index 632 L  \n",
      "Index 633 F  \n",
      "Index 634 K  \n",
      "Index 635 R  \n",
      "Index 636 M  \n",
      "Index 637 I  \n",
      "Index 638 D  \n",
      "Index 639 L  \n",
      "Index 640 S  \n",
      "Index 641 S  \n",
      "Index 642 I  \n",
      "Index 643 S  \n",
      "Index 644 T  \n",
      "Index 645 V  \n",
      "Index 646 D  \n",
      "Index 647 S  \n",
      "Index 648 M  \n",
      "Index 649 I  \n",
      "Index 650 A  \n",
      "Index 651 L  \n",
      "Index 652 D  \n",
      "Index 653 C  \n",
      "Index 654 D  \n",
      "Index 655 P  \n",
      "Index 656 L  \n",
      "Index 657 C  \n",
      "Index 658 N  \n",
      "Index 659 T  \n",
      "Index 660 D  \n",
      "Index 661 F  \n",
      "Index 662 R  \n",
      "Index 663 V  \n",
      "Index 664 L  \n",
      "Index 665 E  \n",
      "Index 666 L  \n",
      "Index 667 Y  \n",
      "Index 668 S  \n",
      "Index 669 Q  \n",
      "Index 670 K  \n",
      "Index 671 E  \n",
      "Index 672 L  \n",
      "Index 673 R  \n",
      "Index 674 S  \n",
      "Index 675 S  \n",
      "Index 676 N  \n",
      "Index 677 V  \n",
      "Index 678 F  \n",
      "Index 679 D  \n",
      "Index 680 L  \n",
      "Index 681 E  \n",
      "Index 682 E  \n",
      "Index 683 I  \n",
      "Index 684 M  \n",
      "Index 685 R  \n",
      "Index 686 E  \n",
      "Index 687 F  \n",
      "Index 688 N  \n",
      "Index 689 S  \n",
      "Index 690 Y  \n",
      "Index 691 K  \n",
      "Index 692 Q  \n",
      "Index 693 R  \n",
      "Index 694 V  \n",
      "Index 695 K  \n",
      "Index 696 Y  \n",
      "Index 697 V  \n",
      "Index 698 E  \n",
      "Index 699 D  \n",
      "Index 700 K  \n",
      "Index 701 V  \n",
      "Index 702 V  \n",
      "Index 703 D  \n",
      "Index 704 P  \n",
      "Index 705 G  \n",
      "Index 706 S  \n",
      "Index 707 G  \n",
      "Index 708 Y  \n",
      "Index 709 I  \n",
      "Index 710 P  \n",
      "Index 711 E  \n",
      "Index 712 A  \n",
      "Index 713 P  \n",
      "Index 714 R  \n",
      "Index 715 D  \n",
      "Index 716 G  \n",
      "Index 717 Q  \n",
      "Index 718 A  \n",
      "Index 719 Y  \n",
      "Index 720 V  \n",
      "Index 721 R  \n",
      "Index 722 K  \n",
      "Index 723 D  \n",
      "Index 724 G  \n",
      "Index 725 E  \n",
      "Index 726 W  \n",
      "Index 727 V  \n",
      "Index 728 L  \n",
      "Index 729 L  \n",
      "Index 730 S  \n",
      "Index 731 T  \n",
      "Index 732 F  \n",
      "Index 733 L  \n",
      "Index 734 G  \n",
      "Index 735 A  \n",
      "Index 736 A  \n",
      "Index 737 A  \n",
      "Index 738 S  \n",
      "Index 739 L  \n",
      "Index 740 E  \n",
      "Index 741 V  \n",
      "Index 742 L  \n",
      "Index 743 F  \n",
      "Index 744 Q  \n",
      "Index 745 G  \n",
      "Index 746 P  \n",
      "Index 747 G  \n",
      "Index 748 H  \n",
      "Index 749 H  \n",
      "Index 750 H  \n",
      "Index 751 H  \n",
      "Index 752 H  \n",
      "Index 753 H  \n",
      "Index 754 H  \n",
      "Index 755 H  \n",
      "Index 756 S  \n",
      "Index 757 A  \n",
      "Index 758 W  \n",
      "Index 759 S  \n",
      "Index 760 H  \n",
      "Index 761 P  \n",
      "Index 762 Q  \n",
      "Index 763 F  \n",
      "Index 764 E  \n",
      "Index 765 K  \n",
      "Index 766 G  \n",
      "Index 767 G  \n",
      "Index 768 A  \n",
      "Index 769 S  \n",
      "Index 770 G  \n",
      "Index 771 G  \n",
      "Index 772 G  \n",
      "Index 773 G  \n",
      "Index 774 S  \n",
      "Index 775 G  \n",
      "Index 776 G  \n",
      "Index 777 S  \n",
      "Index 778 A  \n",
      "Index 779 W  \n",
      "Index 780 S  \n",
      "Index 781 H  \n",
      "Index 782 P  \n",
      "Index 783 Q  \n",
      "Index 784 F  \n",
      "Index 785 E  \n",
      "Index 786 K  \n"
     ]
    }
   ],
   "source": [
    "# Debug to check alignment and cutting is correct\n",
    "for i in range(np.size(seq_replaced)):\n",
    "    print(\"Index\", i + 1 + pdb_offset, seq_replaced[i], conservative_seq[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score Ratio 0.0\n"
     ]
    }
   ],
   "source": [
    "# Report in text\n",
    "current_step = []\n",
    "tot_size = 0\n",
    "\n",
    "for i in range(np.size(conservative_seq)):\n",
    "    if conservative_seq[i] == \"*\":\n",
    "        current_step.append(i + 1)\n",
    "\n",
    "    elif conservative_seq[i] == \":\":\n",
    "        current_step.append(i + 1)\n",
    "\n",
    "    elif np.size(current_step) >= min_seq_len_to_report:\n",
    "        print(\"Found conservation region at amino position {0} - {1}, with length {2}\".format(current_step[0] + pdb_offset, current_step[-1] + pdb_offset, np.size(current_step)))\n",
    "        tot_size = tot_size + np.size(current_step)\n",
    "        current_step = []\n",
    "\n",
    "    elif np.size(current_step) < min_seq_len_to_report:\n",
    "        tot_size = tot_size + np.size(current_step)\n",
    "        current_step = []\n",
    "    \n",
    "    else:\n",
    "        pass\n",
    "\n",
    "try:\n",
    "    print(\"Found conservation region at amino position {0} - {1}, with length {2}\".format(current_step[0] + pdb_offset, current_step[-1] + pdb_offset, np.size(current_step)))\n",
    "except:\n",
    "    pass\n",
    "\n",
    "print(\"Score Ratio\", np.round(tot_size / seq_len, 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "select /A, B, C; color sel dark grey\n",
      "preset \"overall look\" \"publication 1 (silhouettes)\"\n",
      "save 5_chimera.cxs\n"
     ]
    }
   ],
   "source": [
    "# Report in Command\n",
    "current_step = []\n",
    "print(\"select /{1}; color sel {0}\".format(chain_color, chain))\n",
    "\n",
    "for i in range(np.size(conservative_seq)):\n",
    "    if conservative_seq[i] == \"*\":\n",
    "        current_step.append(i + 1)\n",
    "\n",
    "    elif conservative_seq[i] == \":\":\n",
    "        current_step.append(i + 1)\n",
    "\n",
    "    elif conservative_seq[i] == \".\":\n",
    "        current_step.append(i + 1)\n",
    "\n",
    "    elif np.size(current_step) >= min_seq_len_to_report:\n",
    "        print(\"select /{0}:{1}-{2}; color sel {3}\".format(chain, current_step[0] + pdb_offset, current_step[-1] + pdb_offset, color_list[i%np.size(color_list)]))\n",
    "        current_step = []\n",
    "\n",
    "    elif np.size(current_step) < min_seq_len_to_report:\n",
    "        current_step = []\n",
    "    \n",
    "    else:\n",
    "        pass\n",
    "\n",
    "try:\n",
    "    print(\"select /{0}:{1}-{2}; color sel {3}\".format(chain, current_step[0] + pdb_offset, current_step[-1] + pdb_offset, color_list[i%np.size(color_list)]))\n",
    "except:\n",
    "    pass\n",
    "\n",
    "print('preset \"{0}\" \"{1}\"'.format(\"overall look\", \"publication 1 (silhouettes)\"))\n",
    "print(\"save 5_chimera.cxs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOTE: available bundle cache has not been initialized yet\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    !chimerax;\n",
    "except:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
